src/controller/rec_controller.c

```diff
@@ -190,11 +201,56 @@ int rec_controller_poll(controller *ctrl, ctrl_event **ev) {
     return 0;
 }
 
+int rec_controller_tick(controller *ctrl, uint32_t ticks0, ctrl_event **ev) {
+    wtf *data = ctrl->data;
+    if(data->id == 0 && ticks0 % 10 == 0) {
+        if(scene_is_arena(game_state_get_scene(ctrl->gs)) &&
+           game_state_find_object(ctrl->gs, game_player_get_har_obj_id(game_state_get_player(ctrl->gs, 1)))) {
+            game_state *gs_bak = vector_append_ptr(&data->game_states);
+            game_state_clone(ctrl->gs, gs_bak);
+        }
+    }
+    return 0;
+}
+
+void rec_controller_step_back(controller *ctrl) {
+    wtf *data = ctrl->data;
+    ctrl->last = ACT_STOP;
+    data->last_action = ACT_STOP;
+    if(data->id != 0) {
+        return;
+    }
+    game_state *gs_bak = vector_back(&data->game_states);
+    while(vector_size(&data->game_states) > 1 && gs_bak->int_tick >= ctrl->gs->int_tick) {
+        game_state_clone_free(gs_bak);
+        vector_pop(&data->game_states);
+        gs_bak = vector_back(&data->game_states);
+    }
+
+    game_state *gs_new = omf_calloc(1, sizeof(game_state));
+    game_state_clone(gs_bak, gs_new);
+    gs_new->clone = false;
+    ctrl->gs->new_state = gs_new;
+
+    log_debug("REWOUND game state from %d to %d", ctrl->gs->int_tick, gs_new->int_tick);
+
+    // fix the game state pointers in the controllers
+    for(int i = 0; i < game_state_num_players(gs_new); i++) {
+        game_player *gp = game_state_get_player(gs_new, i);
+        controller *c = game_player_get_ctrl(gp);
+        if(c) {
+            c->gs = gs_new;
+        }
+    }
+}
+
 void rec_controller_create(controller *ctrl, int player, sd_rec_file *rec) {
     wtf *data = omf_calloc(1, sizeof(wtf));
     data->last_tick = 0;
     data->last_action = ACT_STOP;
+    data->id = player;
```

Comment:
please rename wtf's id field to player_id

Reply:
Done, and I renamed the struct as well.